Fix PPG & VNG/VNG4 demosaicer CPU code

jenshannoschwalm opened the pull request:
Both - CPU & GPU - paths should produce almost identical results. The PPG demosaicer didn't work that way leading to significant differences.

We now make sure output is always at least zero for all channels as we do in OpenCL code

jenshannoschwalm commented:
Some comments about this
1. The "isssue for wrong ppg" corrections is from very old dcraw code, had been corrected for OpenCL because of AMD issues
2. This is responsible for a vast number of cpu vs gpu diffs in integration tests if the used demosaicer is ppg
3. First ever bug detected via evaluation of the --dump-pipe tool :-)

TurboGit commented:
> 3\. First ever bug detected via evaluation of the --dump-pipe tool :-)

Yeah! You put in place quite some good tools for checking the dt internals... and this now pays :)

TurboGit commented:
The good news is that really help having the CPU & GPU path closer.

Before & After for some tests:

```
Test 0000-nop
      Image mire1.cr2
      CPU & GPU version differ by 65131 pixels

Test 0000-nop
      Image mire1.cr2
      CPU & GPU version differ by 8323 pixels
```

```
Test 0062-splittoning
      Image mire1.cr2
      CPU & GPU version differ by 66322 pixels

Test 0062-splittoning
      Image mire1.cr2
      CPU & GPU version differ by 17318 pixels
```

```
Test 0035-multiple-modules
      Image mire1.cr2
      CPU & GPU version differ by 692194 pixels

Test 0035-multiple-modules
      Image mire1.cr2
      CPU & GPU version differ by 663068 pixels
```

The bad news... I'll have to update all the integration tests expected output !

So one question, do you plan having some more fixes in this as I'd like to avoid another full update of the integration test expected output.

jenshannoschwalm commented:
I will have a look tonight on the other demosaicers if there are discrepancies, so ready to merge this but wait for updating the references for a day?

TurboGit commented:
I prefer waiting to avoid having all the tests failing and possibly hiding an issue introduced by another PR. There is no urgency, so just le me know when you've analyzed the other demosaicers.

TurboGit commented:
Also please add entry for the release notes. TIA.

jenshannoschwalm commented:
1. Spotted diffs and fixed for green_equilibration
2. Spotted problems for vng4 both in cpu & gpu code. Now identical results.

So nothing more from me on this topic.

Release notes would be like:
- Fixed CPU vs OpenCl output differences in PPG and VNG/VNG4 demosaicers, match greens and color smoothing.

Remaining questions from me as comments in https://github.com/darktable-org/darktable-tests/pull/14

ralfbrown commented on src/iop/demosaic.c:
```diff
@@ -2314,6 +2314,14 @@ static void lin_interpolate(
    I've extended the basic idea to work with non-Bayer filter arrays.
    Gradients are numbered clockwise from NW=0 to W=7.
 */
+static inline void _ensure_abovezero(void *to, void *from, int num)
```
Is there any reason for using `void*` parameters instead of `float*` ?  It looks like all calls are using float* variables anyway.

jenshannoschwalm replied:
Confessing I don't know a better way. @TurboGit do you have a proper way?

TurboGit replied:
Maybe doing explicit float* cast:

```
diff --git a/src/iop/demosaic.c b/src/iop/demosaic.c
index b1e501f25d..159114e540 100644
--- a/src/iop/demosaic.c
+++ b/src/iop/demosaic.c
@@ -2314,10 +2314,8 @@ static void lin_interpolate(
    I've extended the basic idea to work with non-Bayer filter arrays.
    Gradients are numbered clockwise from NW=0 to W=7.
 */
-static inline void _ensure_abovezero(void *to, void *from, int num)
+static inline void _ensure_abovezero(float *out, float *in, const int num)
 {
-  float *in = from;
-  float *out = to;
   for(int i=0; i < num; i++)
     out[i] = fmaxf(0.0f, in[i]);
 }
@@ -2483,13 +2481,19 @@ static void vng_interpolate(
       }
     }
     if(row > 3) /* Write buffer to image */
-      _ensure_abovezero(out + 4 * ((row - 2) * width + 2), brow[0] + 2, 4 * (width - 4));
+      _ensure_abovezero((float *)(out + 4 * ((row - 2) * width + 2)),
+                        (float *)(brow[0] + 2),
+                        4 * (width - 4));
     // rotate ring buffer
     for(int g = 0; g < 4; g++) brow[(g - 1) & 3] = brow[g];
   }
   // copy the final two rows to the image
-  _ensure_abovezero(out + (4 * ((height - 4) * width + 2)), brow[0] + 2, 4 * (width - 4));
-  _ensure_abovezero(out + (4 * ((height - 3) * width + 2)), brow[1] + 2, 4 * (width - 4));
+  _ensure_abovezero((float *)(out + (4 * ((height - 4) * width + 2))),
+                    (float *)(brow[0] + 2),
+                    4 * (width - 4));
+  _ensure_abovezero((float *)(out + (4 * ((height - 3) * width + 2))),
+                    (float *)(brow[1] + 2),
+                    4 * (width - 4));
   dt_free_align(buffer);
 
   if(filters != 9 && !FILTERS_ARE_4BAYER(filters)) // x-trans or CYGM/RGBE
```

ralfbrown replied:
I'm surprised *any* cast would be needed - `out` and `brow[i]` are both `float*` and adding an integer value to a `float*` doesn't change type.

TurboGit replied:
It is needed because brow[1] is is float[4].

jenshannoschwalm replied:
I am/was surprised too but thought my C knowledge is too low. I just didn't compile.

parafin replied:
smth like `&(brow[0] + 2)[0]` might work, though not sure if better than casting

jenshannoschwalm replied:
Friends, I always thought FORTH is the programming language with the best chance to do things correctly but without a chance to understand :-) I'd call this obfuscating :-)

TurboGit replied:
@jenshannoschwalm : Yes C is showing its age... It's a bad language compared to what we have today in terms of readability.

TurboGit replied:
I don't like it as-is as the spec of  `_ensure_abovezero` is really misleading with the `void *`.

victoryforce replied:
> Friends, I always thought FORTH is the programming language with the best chance to do things correctly but without a chance to understand :-)

As someone who wrote a lot in FORTH more than 30 years ago, I would say that you can write super-understandable programs in this language. Although it is just as easy to write a program that is understandable only to the author. Everything is like in C, overall. 😉

jenshannoschwalm replied:
Fully agreed Victor! I mostly did FORTH for more than 30yrs and just got myself into C because of darktable and the people here ! So far i am reluctant to integrate some FORTH into dt, would definitely prefer that over lua or bash :-)

ralfbrown replied:
Dabbled in FORTH around the time Victor was using it.  Has some fun attributes such as being able to change the syntax of the language or (like Lisp) being able to write a FORTH system in FORTH once you implement a handfull of small core functions in assembly.  I came across some blog posts about implementing a FORTH in a bootsector a couple of years ago: https://compilercrim.es/bootstrap/

victoryforce replied:
> I came across some blog posts about implementing a FORTH in a bootsector a couple of years ago: https://compilercrim.es/bootstrap/

It is interesting that my main projects at that time were related to the implementation of Forth on "bare hardware". One of the projects was the Forth-bootstrapping BIOS for a VLIW-based computer (used in the systems of the Soviet Navy). After the Soviet Union died (thank God!), I wrote a part of the Forth system, which was loaded from the BIOS directly or, alternatively, from a magnetic tape, or from a floppy disk for a not fully IBM PC compatible Ukrainian computer. I still remember that the graphics editor code that was part of that system took up 4KB of RAM.

jenshannoschwalm commented:
Now ready from my side ...

TurboGit reviewed (approved):
Thanks!
